cosi/protocol/protocol.go

```diff
@@ -0,0 +1,354 @@
+	onet.GlobalProtocolRegister(DefaultSubProtocolName, NewDefaultSubProtocol)
+}
+
+// NewProtocol method is used to define the protocol.
+func NewProtocol(n *onet.TreeNodeInstance, vf VerificationFn, subProtocolName string, suite cosi.Suite) (onet.ProtocolInstance, error) {
+
+	var list []kyber.Point
+	for _, t := range n.Tree().List() {
+		list = append(list, t.ServerIdentity.Public)
+	}
+
+	c := &CoSiRootNode{
+		TreeNodeInstance: n,
+		FinalSignature:   make(chan []byte, 1),
+		Data:             make([]byte, 0),
+		publics:          list,
+		hasStopped:       false,
+		startChan:        make(chan bool, 1),
+		verificationFn:   vf,
+		subProtocolName:  subProtocolName,
+		suite:            suite,
+	}
+
+	return c, nil
+}
+
+// Shutdown stops the protocol
+func (p *CoSiRootNode) Shutdown() error {
+	if !p.hasStopped {
+		p.hasStopped = true
+	}
+	return nil
+}
+
+// Dispatch is the main method of the protocol, defining the root node behaviour
+// and sequential handling of subprotocols.
+func (p *CoSiRootNode) Dispatch() error {
+	if !p.IsRoot() {
+		return nil
+	}
+
+	select {
+	case _, ok := <-p.startChan:
+		if !ok {
+			log.Lvl1("protocol finished prematurely")
+			return nil
+		}
+		close(p.startChan)
+	case <-time.After(time.Second):
+		return fmt.Errorf("timeout, did you forget to call Start?")
+	}
+
+	log.Lvl1("leader protocol started")
+
+	verifyChan := make(chan bool, 1)
+	go func() {
+		log.Lvl3(p.ServerIdentity().Address, "starting verification")
+		verifyChan <- p.verificationFn(p.Msg, p.Data)
+	}()
+
+	// generate trees
+	nNodes := p.Tree().Size()
+	trees, err := genTrees(p.Tree().Roster, nNodes, p.NSubtrees)
+	if err != nil {
+		return fmt.Errorf("error in tree generation: %s", err)
+	}
+
+	// if one node, sign without subprotocols
+	if nNodes == 1 {
+		trees = make([]*onet.Tree, 0)
+	}
+
+	// start all subprotocols
+	cosiSubProtocols := make([]*CoSiSubProtocolNode, len(trees))
+	for i, tree := range trees {
+		cosiSubProtocols[i], err = p.startSubProtocol(tree)
+		if err != nil {
+			return err
+		}
+	}
+	log.Lvl3(p.ServerIdentity().Address, "all protocols started")
+
+	commitments, runningSubProtocols, err := p.collectCommitments(trees, cosiSubProtocols)
+	if err != nil {
+		return err
+	}
+
+	// generate challenge
+	log.Lvl3("root-node generating global challenge")
+	secret, commitment, finalMask, err := generateCommitmentAndAggregate(p.suite, p.TreeNodeInstance, p.publics, commitments)
+	if err != nil {
+		return err
+	}
+
+	cosiChallenge, err := cosi.Challenge(p.suite, commitment, finalMask.AggregatePublic, p.Msg)
+	if err != nil {
+		return err
+	}
+	structChallenge := StructChallenge{p.TreeNode(), Challenge{cosiChallenge}}
+
+	// send challenge to every subprotocol
+	for _, coSiProtocol := range runningSubProtocols {
+		subProtocol := coSiProtocol
+		subProtocol.ChannelChallenge <- structChallenge
+	}
+
+	// get response from all subprotocols
+	responses := make([]StructResponse, 0)
+	errChan := make(chan error, 1)
```

Comment:
You are right, we should not swallow errors which may make debugging more difficult later on. I did it this way is because the logic is simpler when I need to read the channel.